rust/chains/tw_pactus/src/transaction.rs

```diff
@@ -0,0 +1,410 @@
+    Bond = 2,
+    Sortition = 3,
+    Unbond = 4,
+    Withdraw = 5,
+}
+
+impl TryFrom<u8> for PayloadType {
+    type Error = EncoderError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(PayloadType::Transfer),
+            2 => Ok(PayloadType::Bond),
+            3 => Ok(PayloadType::Sortition),
+            4 => Ok(PayloadType::Unbond),
+            5 => Ok(PayloadType::Withdraw),
+            _ => Err(EncoderError::ParseFailed("Invalid PayloadType value")),
+        }
+    }
+}
+
+impl Encodable for PayloadType {
+    fn encode(&self, w: &mut dyn std::io::Write) -> Result<(), EncoderError> {
+        (self.clone() as u8).encode(w)
+    }
+
+    fn encoded_size(&self) -> usize {
+        1
+    }
+}
+
+impl Decodable for PayloadType {
+    fn decode(r: &mut dyn std::io::Read) -> Result<Self, EncoderError> {
+        PayloadType::try_from(u8::decode(r)?)
+    }
+}
+
+pub trait Payload: Debug + Encodable {
+    fn signer(&self) -> &Address;
+    fn value(&self) -> Amount;
+    fn payload_type(&self) -> PayloadType;
+}
+
+#[derive(Debug)]
+pub struct TransferPayload {
+    sender: Address,
+    receiver: Address,
+    amount: Amount,
+}
+
+impl Encodable for TransferPayload {
+    fn encode(&self, w: &mut dyn std::io::Write) -> Result<(), EncoderError> {
+        self.sender.encode(w)?;
+        self.receiver.encode(w)?;
+        self.amount.encode(w)?;
+
+        Ok(())
+    }
+
+    fn encoded_size(&self) -> usize {
+        self.sender.encoded_size() + self.receiver.encoded_size() + self.amount.encoded_size()
+    }
+}
+
+impl Decodable for TransferPayload {
+    fn decode(r: &mut dyn std::io::Read) -> Result<Self, EncoderError> {
+        let sender = Address::decode(r)?;
+        let receiver = Address::decode(r)?;
+        let amount = Amount::decode(r)?;
+
+        Ok(TransferPayload {
+            sender,
+            receiver,
+            amount,
+        })
+    }
+}
+
+impl Payload for TransferPayload {
+    fn signer(&self) -> &Address {
+        &self.sender
+    }
+    fn value(&self) -> Amount {
+        self.amount.clone()
+    }
+    fn payload_type(&self) -> PayloadType {
+        PayloadType::Transfer
+    }
+}
+
+#[derive(Debug)]
+pub struct Transaction {
+    flags: u8,
+    version: u8,
+    lock_time: u32,
+    fee: Amount,
+    memo: String,
+    payload: Box<dyn Payload>,
+    signature: Option<Signature>,
+    public_key: Option<PublicKey>,
+}
+
+impl Transaction {
+    pub fn new(
+        flags: u8,
+        version: u8,
+        lock_time: u32,
+        fee: Amount,
+        memo: String,
+        payload: Box<dyn Payload>,
+    ) -> Self {
+        Transaction {
+            flags,
+            version,
+            lock_time,
+            fee,
+            memo,
+            payload,
+            public_key: None,
+            signature: None,
+        }
+    }
+
+    pub fn from_proto(input: &Pactus::Proto::SigningInput) -> SigningResult<Self> {
```

Comment:
Could you please move `Transaction::from_proto` implementation to a separate module.
We try to separate TrustWallet's Protobuf interface and Rust entities.
For example, take a look at `TxBuilder` for BNB chain:
https://github.com/trustwallet/wallet-core/blob/406abe4b0649a74279b2e7b77a2b10d1bb61520c/rust/chains/tw_binance/src/modules/tx_builder.rs#L12-L29

Reply:
It is really good idea to separate them. 
I can move it to `transaction_util.rs` if you want.